internal/services/elastic/elastic.go

```diff
@@ -0,0 +1,232 @@
+	}
+}
+
+// buildHistoryQuery builds the query to retrieve the history of a device from elastic.
+func buildHistoryQuery(params GetHistoryParams) search.Request {
+	// convert the start and end times to milliseconds since epoch
+	startArg := strconv.Itoa(int(params.StartTime.UnixMilli()))
+	endArg := strconv.Itoa(int(params.EndTime.UnixMilli()))
+
+	// calculate the interval in milliseconds
+	interval := calculateInterval(params.StartTime, params.EndTime, params.Buckets)
+	intervalArg := strconv.Itoa(interval) + "ms"
+
+	msm := types.MinimumShouldMatch(1)
+	min := 1
+	filter := getHistoricalPrivilegeFilter(params.PrivilegeIDs)
+	return search.Request{
+		// get documents that match the device ID, are within the time range, and have a vehicle field
+		Query: &types.Query{
+			Bool: &types.BoolQuery{
+				Filter: []types.Query{
+					{Term: map[string]types.TermQuery{subjectField: {Value: params.DeviceID}}},
+					{Range: map[string]types.RangeQuery{timeField: types.DateRangeQuery{Gte: &startArg, Lte: &endArg}}},
+				},
+				Should: []types.Query{
+					{Exists: &types.ExistsQuery{Field: "data.vehicle"}},
+				},
+				MinimumShouldMatch: &msm,
+			},
+		},
+		Aggregations: map[string]types.Aggregations{
+			// docByInterval histogram aggregates the data into buckets based on the interval
+			docByInterval: {
+				DateHistogram: &types.DateHistogramAggregation{
+					Field:         ptr(timeField),
+					FixedInterval: &intervalArg,
+					MinDocCount:   &min,
+				},
+				Aggregations: map[string]types.Aggregations{
+					// singleDoc aggregation selects the first data point in each bucket
+					// that matches the desired filter
+					singleDoc: {
+						TopHits: &types.TopHitsAggregation{
+							Size:    ptr(1),
+							Source_: &filter,
+						},
+					},
+				},
+			},
+		},
+		Size: ptr(0), // only return the aggregations
+	}
+}
+
+// calculateInterval calculates the desired interval in milliseconds. This interval will provide n buckets between the start and end times.
+// if the interval is 0 or negative it will be clamped to 1ms.
+func calculateInterval(start, end time.Time, buckets int) int {
+	duration := int(end.Sub(start).Milliseconds())
+	interval := duration / buckets
+	if interval < 1 {
+		return 1 // 0 or negative interval is not supported, clamp to 1
+	}
+	return interval
+}
+
+// getHitsFromHistoryResponse extracts the hits from the response body.
+// The maxValues parameter is used to limit the number of values returned.
+func getHitsFromHistoryResponse(body []byte, maxValues int) []any {
+	retData := []any{}
+	bucketKey := fmt.Sprintf("aggregations.%s.buckets", docByInterval)
+	hitsKey := fmt.Sprintf("%s.hits.hits", singleDoc)
+	gjson.GetBytes(body, bucketKey).ForEach(func(key, v gjson.Result) bool {
+		gjson.Get(v.Raw, hitsKey).ForEach(func(key, value gjson.Result) bool {
+			// if we have reached the max number of values, stop
+			if len(retData) == maxValues {
```

Comment:
How would we go above this?

Reply:
Ah, when we round down when we calculate the interval.

Reply:
It has to do with the exact time your start and end fall on and how elastic does rounding for it's aggregation.

If you look through the test cases I added one called "buckets + 1 returned from elastic" and you can see that those exact start and stop time will cause elastic to provide an extra bucket.

Formula and docs on how bucket key is determined.
[Docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-datehistogram-aggregation.html)
> bucket_key = Math.floor(time / interval) * interval

Reply:
That makes sense to me. Maybe leave a comment? I know tests are documentation and blah blah.